### PR TITLE
KIALI-1528 Introduce new Overview page

### PR DESCRIPTION
**Pull request by jotak**
- Navigation / menu entry
- Namespace cards
- Fetching backend app health
- Links to other sections with pre-selected namespace

JIRA: https://issues.jboss.org/browse/KIALI-1528

Still to do:

- [x] Options bar
- [x] Tests

### CODE REVIEW

**Comment by jotak** [comment]
PR opened so that I can get early feedback.

Screenshot:
cards

Screenshot with popover:
popover

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-ui-pr-e2e-test [#15](https://paste.fedoraproject.org/paste/a0qri-vncb-TtNmSvaCjdg)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1177]](https://paste.fedoraproject.org/paste/TB3ElkquyimqgBlOh13A9g)
  * :heavy_check_mark: Kiali-on-Openshift [#[527]](https://paste.fedoraproject.org/paste/fTJh6vuCtdrZBirD8dmNUw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[943]](https://paste.fedoraproject.org/paste/83FpO4MXHjek9RZEbH6JLg)
  * :red_circle: run-kiali-e2e-tests [#[28]](https://paste.fedoraproject.org/paste/K~mbHkd6R6I7nZagn6GIhg)

**Comment by israel-hdez** [comment]
Looks nice.
The icons at the bottom of the cards link, are they links? If they are, it's not very intuitive. But I don't have a better recommendation other than suggesting to place a label. That will make the cards less clean :( So, i would vote for keeping them as in your current proposal (duh!)

**Comment by pilhuhn** [comment]
Looks great!

@jotak Can you  make this available in master under the experimental flag?
This way one can play with it, which is more intuitive for feedback than screenshots.

**Comment by abonas** [comment]
@cshinn @serenamarie125 please review.
@jotak this is very nice :)
I think that the cards where all is green have some duplication - 16 applications and then under it again 16 with a green mark. Perhaps both rows could be consolidated?
Also, the X applications should be clickable (and so should be the Y apps in error) and navigate to the apps list page with the appropriate filter. IIRC it was part of the design spec, but I could be wrong.

**Comment by jotak** [comment]
@abonas yes the statuses are not linking to filtered lists yet. I think I'll enter another JIRA for that (hence for another PR), as the list pages are not yet able to filter by health

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-ui-pr-e2e-test [#41](https://paste.fedoraproject.org/paste/kpxQ1HOg1n~d7bwDzMiBlQ)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1242]](https://paste.fedoraproject.org/paste/0y683zfNbdhcpd2zAAYZxA)
  * :heavy_check_mark: Kiali-on-Openshift [#[587]](https://paste.fedoraproject.org/paste/R3RDd-A4~h5W8QJsMkjvxA)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1015]](https://paste.fedoraproject.org/paste/HSPwDrjc9Kyqpz0g~83RgA)
  * :red_circle: run-kiali-e2e-tests [#[63]](https://paste.fedoraproject.org/paste/ee3kmI6saMe0a35v~BM4jA)


KIALI_NAME=kialiuipr662-istio-system

**Comment by abonas** [comment]
@jotak have you changed anything visual in the last commit? is it reflected in the screenshots? thank you

**Comment by abonas** [comment]
> @abonas yes the statuses are not linking to filtered lists yet. I think I'll enter another JIRA for that (hence for another PR), as the list pages are not yet able to filter by health

do you have the JIRA #?

**Comment by jotak** [comment]
New visual, now including the full options bar:
overview

"N Applications" text is now a link.

**Comment by jotak** [comment]
@abonas https://issues.jboss.org/browse/KIALI-1560 and https://issues.jboss.org/browse/KIALI-1561

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-ui-pr-e2e-test [#55](https://paste.fedoraproject.org/paste/7-qTvokqh-hLKTUFouVrNA)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1275]](https://paste.fedoraproject.org/paste/ySz-0xsNfLKkv6Sq-AGeZw)
  * :heavy_check_mark: Istio-Upstream [#[160]](https://paste.fedoraproject.org/paste/XGLeF9Z~Q-EM7GaRSjGmQw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1066, 1067]](https://paste.fedoraproject.org/paste/wWztJdUj-mbeJAw8e7BdEw)
  * :heavy_check_mark: install-bookinfo [#[625]](https://paste.fedoraproject.org/paste/sqdNRimVEMuuRCSc6aGXww)
  * :heavy_check_mark: Kiali-on-Openshift [#[622]](https://paste.fedoraproject.org/paste/UhpmzrAbSkry8w~PXNWtNg)
  * :heavy_check_mark: run-kiali-e2e-tests [#[92]](https://paste.fedoraproject.org/paste/ffhbxRyksbFRtDz3nwxDaA)


KIALI_NAME=kialiuipr662-istio-system

**Comment by abonas** [comment]
@jotak looks great! 
Visually wise I recommend to merge. the rest of the enhancements can be added later on top.

**Comment by jotak** [comment]
Thanks! I'm keeping the DNM label just because this PR is based on a refactoring that is not merged yet (#667) so just a matter of synchronizing PRs. Otherwise it's ready for review.

**Comment by jshaughn** [comment]
I'm not sure if this is part of the design but being able to select multiple namespaces here would be good.  Then you could navigate to a multi-namespace graph or have a pre-set namespaces filter.

**Comment by abonas** [comment]
> I'm not sure if this is part of the design but being able to select multiple namespaces here would be good. Then you could navigate to a multi-namespace graph or have a pre-set namespaces filter.

it is not part of the design, and it is better to build those capabilities step by step and not over complicate it here.

**Comment by pilhuhn** [comment]
@jshaughn As @abonas said, lets get this in and then add on top. This way it is easier to give feedback.

**Comment by jshaughn** [comment]
It wasn't my intent that it be added.  I was just mentioning it.

**Comment by jmazzitelli** [comment]
Get we get this merged sooner rather than later - I really, really, really hate this "all" namespace view, especially when it is the default.

**Comment by pilhuhn** [comment]
Yes, please merge. If needed with experimental flag.

**Comment by jotak** [comment]
@pilhuhn @jmazzitelli the problem right now is not about experimental flag, it's dependency to another PR which turned out to introduce an unfortunate regression... The current version ( #680 ) seems much better now and I think we're close to merge both PRs.

**Comment by jotak** [comment]
(rebased)
So, the dependent PR is now merged.
But still have a problem: conflicting with #678

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-ui-pr-e2e-test [#77](https://paste.fedoraproject.org/paste/6X-bH-MW4196Ny02rFvBUw)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1310]](https://paste.fedoraproject.org/paste/8qoyE~cAtp6oKmr7lpbl5g)
  * :heavy_check_mark: Istio-Upstream [#[193]](https://paste.fedoraproject.org/paste/q1XBD7746ECUGtYSOkr9Uw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1146, 1147]](https://paste.fedoraproject.org/paste/hhwSUmlvXmceAIPXPwVENQ)
  * :heavy_check_mark: install-bookinfo [#[661]](https://paste.fedoraproject.org/paste/wvPqs9DyJZMvtgNXfRreZg)
  * :heavy_check_mark: Kiali-on-Openshift [#[675]](https://paste.fedoraproject.org/paste/DWGpaEKxW~BE9AYWsji9SQ)
  * :heavy_check_mark: run-kiali-e2e-tests [#[125]](https://paste.fedoraproject.org/paste/HWlGZkEEfsgVBmxpxb5sZQ)


KIALI_NAME=kialiuipr662-istio-system

**Comment by rhqci** [comment]
!!! Couldn't read commit file !!!

**Comment by jotak** [comment]
Please review.
I removed the conflicting parts with #678

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-ui-pr-e2e-test [#80](https://paste.fedoraproject.org/paste/3Q-je1ROHC1fnki3CYb-OA)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1312]](https://paste.fedoraproject.org/paste/DjJz9F~2k5hTXAG-MBo88g)
  * :heavy_check_mark: Istio-Upstream [#[195]](https://paste.fedoraproject.org/paste/l16Nwcl8stqtV3VCJU7p-A)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1152, 1151]](https://paste.fedoraproject.org/paste/kkxL4sMWAtDAVPnqkihidw)
  * :heavy_check_mark: install-bookinfo [#[663]](https://paste.fedoraproject.org/paste/wMUSDX6gKz5i4fgvVoc6ZA)
  * :heavy_check_mark: Kiali-on-Openshift [#[679]](https://paste.fedoraproject.org/paste/jBSoyBlb8DtZL05yHQI~xg)
  * :heavy_check_mark: run-kiali-e2e-tests [#[127]](https://paste.fedoraproject.org/paste/-KUt8-5kRbGj7sYhh66FDg)


KIALI_NAME=kialiuipr662-istio-system

**Comment by cshinn** [comment]
@jotak Looks good! The shortcuts are clever. I think the page would look better if the grey background extended to the sides and bottom of the page below the toolbar

File: src/components/Refresh/Refresh.tsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  onSelect: (selected: PollIntervalInMs) => void;
+};
+
+type State = {
+  pollInterval?: PollIntervalInMs;
+  pollerRef?: number;
+};
+
+class Refresh extends React.Component<Props, State> {
+  static readonly POLL_INTERVALS = config().toolbar.pollInterval;
```
Comment (lucasponce) [inline src/components/Refresh/Refresh.tsx]: Just my curiosity, is better to use "static readonly" vs "const" on these use cases ?

File: src/components/Refresh/Refresh.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                  active={key === this.state.pollInterval}
+                  onSelect={this.updatePollInterval}
+                >
+                  {Refresh.POLL_INTERVALS[key]}
+                </MenuItem>
+              );
+            })}
+          </DropdownButton>
+          <span style={{ paddingLeft: '0.5em' }}>
+            <Button id="refresh_button" onClick={this.props.handleRefresh}>
```
Comment (lucasponce) [inline src/components/Refresh/Refresh.tsx]: Setting an "id" in the component, perhaps could duplicates with the same component in other page ?
In other words, can the component be used in more areas without having this id="refresh_button" defined multiple times ?

File: src/pages/Overview/FiltersAndSorts.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    id: string;
+    title: string;
+    isNumeric: boolean;
+    param: string;
+    compare: (a: NamespaceInfo, b: NamespaceInfo) => number;
+  }
+
+  export const sortFields: SortField[] = [
+    {
+      id: 'namespace',
```
Comment (lucasponce) [inline src/pages/Overview/FiltersAndSorts.ts]: id is 'namespace' but title is 'Name' is this on purpose ?
I guess this is to sort by name, isn't it ?

File: src/pages/Overview/OverviewPage.tsx [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+  componentDidMount() {
+    this.load();
+  }
+
+  load = () => {
+    API.getNamespaces(authentication())
+      .then(namespacesResponse => {
+        const namespaces: Namespace[] = namespacesResponse['data'];
+        this.fetchAppsHealth(namespaces.map(namespace => namespace.name), this.currentDuration());
```
Comment (lucasponce) [inline src/pages/Overview/OverviewPage.tsx]: This is going to filter all namespaces even if filtered.
Or perhaps I missed some part in the code ?

File: src/components/Refresh/Refresh.tsx (hunk 2 above)
Comment (jotak) [reply src/components/Refresh/Refresh.tsx]: You're right, I'll change it.

File: src/pages/Overview/FiltersAndSorts.ts (hunk 3 above)
Comment (jotak) [reply src/pages/Overview/FiltersAndSorts.ts]: Yes by name, but it's the namespace name. I find "namespace" more accurate for an id. But it's displaying "Name" as per UXD mockup.

File: src/pages/Overview/OverviewPage.tsx (hunk 4 above)
Comment (jotak) [reply src/pages/Overview/OverviewPage.tsx]: Yes, the filter doesn't expect exact matching so I have to get all namespace names before filtering.
However there's indeed some place for improvement, as we could perform the namespace filtering _before_ calling `fetchAppHealth`. I'm going to do that.

File: src/components/Refresh/Refresh.tsx (hunk 2 above)
Comment (jotak) [reply src/components/Refresh/Refresh.tsx]: done

File: src/pages/Overview/OverviewPage.tsx (hunk 4 above)
Comment (jotak) [reply src/pages/Overview/OverviewPage.tsx]: Done, improved loading sequence.
That makes me have to rewrite tests

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-ui-pr-e2e-test [#88](https://paste.fedoraproject.org/paste/xPlLwHCdWMUVosC70RoNQQ)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1326]](https://paste.fedoraproject.org/paste/mjycYiSZhBM5XeGSdSlsLQ)
  * :heavy_check_mark: Istio-Upstream [#[207]](https://paste.fedoraproject.org/paste/xZypqDYrvUNTw3d~ZGRICg)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1181, 1182]](https://paste.fedoraproject.org/paste/k5bc7lHn1UMtbe0mHqw0FQ)
  * :heavy_check_mark: install-bookinfo [#[680]](https://paste.fedoraproject.org/paste/48GB8Gsxg39KU8~UufW1XQ)
  * :heavy_check_mark: Kiali-on-Openshift [#[697]](https://paste.fedoraproject.org/paste/b13WN6MKue-eyLhufaezvg)
  * :red_circle: run-kiali-e2e-tests [#[139]](https://paste.fedoraproject.org/paste/PncGKZ6KlamXrvMN~603wA)


KIALI_NAME=kialiuipr662-istio-system

**Comment by jotak** [comment]
@lucasponce could you take another look?
I just rebased (was conflicting), and also fixed an issue with links (see last commit)

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-ui-pr-e2e-test [#95](https://paste.fedoraproject.org/paste/uz2gK~ZR~A8vvnFEEVGE2A)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1356]](https://paste.fedoraproject.org/paste/HbNYoVltThqtO8n~n6iOvQ)
  * :heavy_check_mark: Istio-Upstream [#[235]](https://paste.fedoraproject.org/paste/P1xybGXY2lIvGvMtjNwO~g)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1249, 1250]](https://paste.fedoraproject.org/paste/Uu-BKuo0vrCz6eDO1M46lQ)
  * :heavy_check_mark: install-bookinfo [#[712]](https://paste.fedoraproject.org/paste/DKhynLqeYgqrbrjjuLMUjg)
  * :heavy_check_mark: Kiali-on-Openshift [#[726]](https://paste.fedoraproject.org/paste/n8JNxdW2Fs-sVIxDXuBm8g)
  * :heavy_check_mark: run-kiali-e2e-tests [#[173]](https://paste.fedoraproject.org/paste/kttwn3LpUj5KCvb6w8RNyQ)


KIALI_NAME=kialiuipr662-istio-system

**Comment by lucasponce** [comment]
Sure, sorry, I didn't see your comment.
Reviewieng now.

**Pull request merged**